src/cargo/ops/cargo_rustc/fingerprint.rs

```diff
@@ -454,17 +450,19 @@ pub fn prepare_build_cmd<'a, 'cfg>(cx: &mut Context<'a, 'cfg>, unit: &Unit<'a>)
     // necessary for that fingerprint.
     //
     // Hence, if there were some `rerun-if-changed` directives forcibly change
-    // the kind of fingerprint over to the `MtimeBased` variant where the
-    // relevant mtime is the output path of the build script.
+    // the kind of fingerprint by reinterpreting the dependencies output by the
+    // build script.
     let state = cx.build_state.clone();
     let key = (unit.pkg.package_id().clone(), unit.kind);
+    let root = unit.pkg.root().to_path_buf();
     let write_fingerprint = Work::new(move |_| {
         if let Some(output_path) = output_path {
             let outputs = state.outputs.lock().unwrap();
-            if !outputs[&key].rerun_if_changed.is_empty() {
-                let slot = MtimeSlot(Mutex::new(None));
-                fingerprint.local = LocalFingerprint::MtimeBased(slot,
-                                                                 output_path);
```

Comment:
>I'm also not sure I understand your comment though, can you elaborate?

It was just a gut feeling that something's wrong about this `if`. However today I have a more concrete question :) 

This test currently fails, and I believe that it should pass (according to the current docs):

```Rust
#[test]
fn rerun_if_only_file_changes() {
    let p = project("foo")
        .file("Cargo.toml", r#"
            [package]
            name = "foo"
            version = "0.5.0"
            authors = []
        "#)
        .file("src/main.rs", r#"
            fn main() {
                println!("Hello, World");
            }
        "#)
        .file("build.rs", r#"
            fn main() {
                println!("cargo:rerun-if-env-changed=FOO");
            }
        "#)
        .file("foo", "");
    p.build();

    assert_that(p.cargo("build"),
                execs().with_status(0)
                       .with_stderr("\
[COMPILING] foo v0.5.0 ([..])
[FINISHED] [..]
"));
    sleep_ms(1000);
    File::create(&p.root().join("some-new-file")).unwrap();

    File::create(p.root().join("foo")).unwrap();
    assert_that(p.cargo("build"),
                execs().with_status(0)
                       .with_stderr("\
[COMPILING] foo v0.5.0 ([..])
[FINISHED] [..]
"));
}
```

If `build.rs` does not produce any `rerun-if-changed` then we promise that it will be rerun on *any* change inside the build directory, which is different from the case when `build.rs` specifies an empty set of files in `rerun-if-changed`.

Looks like currently specifying  only `rerun-if-env-changed` implies empty `rerun-if-changed`.